ethcore/src/client/ancient_import.rs

```diff
@@ -53,17 +53,47 @@ impl AncientVerifier {
 		header: &Header,
 		chain: &BlockChain,
 	) -> Result<(), ::error::Error> {
-		match rng.gen::<f32>() <= HEAVY_VERIFY_RATE {
-			true => self.cur_verifier.read().verify_heavy(header)?,
-			false => self.cur_verifier.read().verify_light(header)?,
+		// perform verification
+		let verified = if let Some(ref cur_verifier) = *self.cur_verifier.read() {
+			match rng.gen::<f32>() <= HEAVY_VERIFY_RATE {
+				true => cur_verifier.verify_heavy(header)?,
+				false => cur_verifier.verify_light(header)?,
+			}
+			true
+		} else {
+			false
+		};
+
+		// when there is no verifier initialize it.
+		// We use a bool flag to avoid double locking in the happy case
+		if !verified {
+			let mut cur_verifier = self.cur_verifier.write();
+			if cur_verifier.is_none() {
+				*cur_verifier = Some(self.initial_verifier(header, chain)?);
+			}
+			// Call again to verify.
+			return self.verify(rng, header, chain);
```

Comment:
Couldn't that go in the `else` case branch above directly?

Reply:
Unfortunatelly the lock for `cur_verifier` is still held in `else` branch above, so you cannot lock it again for writing.

But it seems that there is a deadlock here as well, since we call re-cursively `verify` when still holding `cur_verifier` lock, I've updated it.